Support Locale Week Info

JhonnyBillM opened the pull request:
### Summary 

Initial implementation on adding support for Locale Week Info #2999.

Commits are organized similar to the sections in "data pipeline" and "writing a new data struct" docs. 
My current questions and issues are marked with `// TODO: [CODE REVIEW]`.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/unicode-org/icu4x?pullRequest=4171) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/unicode-org/icu4x?pullRequest=4171) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/unicode-org/icu4x?pullRequest=4171) it.</sub>

robertbastian commented on components/locid_transform/src/provider/week_data.rs:
```diff
@@ -0,0 +1,159 @@
+    /// Weekdays values that are part of the 'weekend', for calendar purposes.
+    /// The number of days can be different between locales, and may not be contiguous.
+    #[cfg_attr(feature = "serde", serde(borrow))]
+    pub weekend: ZeroVec<'data, IsoWeekday>,
```
The most efficient way to represent the list of weekdays that make up the weekend would be a bitset on a `u8`. For example `0b00000110` would be the `Saturday, Sunday` weekend, and `0b0000101` the `Friday, Sunday` weekend (we know it's not the `Sunday, Friday` weekend from `first_weekday`).

sffc replied:
+1, I think it's worthwhile doing a more compact representation here since it seems simple enough.

JhonnyBillM replied:
Great suggestion, thank you!
Addressed in new PR #4364. See https://github.com/unicode-org/icu4x/pull/4364/commits/b01b560807915bc8e7a5be73e89a96bcc502e486.

robertbastian reviewed (commented):
We already have the [`icu::calendar::week::WeekCalculator`](https://docs.rs/icu/latest/icu/calendar/week/struct.WeekCalculator.html#structfield.first_weekday) API, which has the `first_weekday` and `min_week_days` fields. It's only missing the weekend information, which could be added with a data struct version bump (i.e. `datetime/week_data@2`) or as a separate data struct.

I'm not sure why @sffc is suggesting that we add this to `icu_locid_transform`, and why there is no mention of `WeekCalculator` in #2999.

sffc reviewed (commented):
Good start!

sffc commented:
> We already have the [`icu::calendar::week::WeekCalculator`](https://docs.rs/icu/latest/icu/calendar/week/struct.WeekCalculator.html#structfield.first_weekday) API, which has the `first_weekday` and `min_week_days` fields. It's only missing the weekend information, which could be added with a data struct version bump (i.e. `datetime/week_data@2`) or as a separate data struct.
> 
> I'm not sure why @sffc is suggesting that we add this to `icu_locid_transform`, and why there is no mention of `WeekCalculator` in #2999.

The proper resolution for `-u-fw` is a somewhat nontrivial algorithm that checks multiple subtags first. See https://unicode-org.atlassian.net/browse/CLDR-16866 (note: this algorithm was just now specified in CLDR, after the ICU4X issue had been filed). I think it likely belongs in `icu_locid_transform`.

I had thought that weekend days was also a Unicode extension keyword but I don't see it in the spec. Still, it is region-based data very similar to `-u-fw` so it seems like it should be co-located.

minDays seems the least related but it is in the same section of both the CLDR spec and the ECMA-402 spec so it seems desirable to keep it co-located. I can also see minDays becoming a more complex resolution algorithm similar to `-u-fw` described above.

Resolution of locale subtags is something that many crates may wish to leverage. When `WeekCalculator` was first written, we did not have a core component for locale operations. Now that we do, I think it makes sense to put code like this in `icu_locid_transform`.

robertbastian commented:
I agree that *resolution* code should be in `icu_locid_transform`, but the actual API (`WeekCalculator`) with fancy types should not. `icu_calendar` can depend on `icu_locid_transform` to check for override tags before loading data. The data whether Saturday is a weekend day in Yemen should not be in `icu_locid_transform`. The data that `-u-fw` can have values `mon`-`sun` can be.

JhonnyBillM commented on components/locid_transform/src/provider/week_data.rs:
```diff
@@ -0,0 +1,159 @@
+/// </div>
+#[icu_provider::data_struct(marker(
+    WeekDataV2Marker,
+    // TODO: [CODE REVIEW]
```
Thanks! Went for V2 in new PR #4364

JhonnyBillM commented:
Thank you folks for the reviews. Addressed most comments in new PR #4364 by defining a new data struct and setting up its data pipeline. Then will open another PR to wire the data to the existing WeekCalculator API.

Unless there is any objection, I will close this PR so we can continue discussions in the new one.

robertbastian commented:
Let's continue in https://github.com/unicode-org/icu4x/pull/4364
